platforms/common/px4_sw_crypto/stub_keystore/stub_keystore.c

```diff
@@ -0,0 +1,57 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2021 Technology Innovation Institute. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+//#include <px4_platform_common/arch_crypto.h>
+#include <string.h>
+#include "public_key.h"
+
+size_t arch_get_key_length(uint8_t idx)
+{
+	return idx < NUMBER_OF_KEYS ? sizeof(public_keys[idx]) : 0;
+}
+
+size_t arch_get_key_by_index(uint8_t idx, uint8_t *key, size_t buf_size)
```

Comment:
Would it be worthwhile to be able to get the key length? I.E. change size_t buf_size to size_t *buf_size, to be able to return the key length in buf_size if *key == null?

Reply:
Yes, you are right. I was only putting in what was needed by the bootloader as it is now. I was planning to address these additional needs in the following pr, which would then add support for doing crypto from px4 modules, such as logger.

Maybe I'll concept those a bit further and add some of that already into this one.

Reply:
I made the following modifications:
- changed the keystore to cache the public and temporary keys in ram, and provide a direct pointer to those. This enables also creating keys in there, and fast access to the keys for crypto algorithms
- added a initializing function, and functions to open/close session to the crypto library. The session is specific to a certain crypto operation/algorithm and can store e.g. nonce values or some HW specific data
- have a separate function for checking the size of the key in keystore. For the key exchange, the crypto libraries need to be able to return encrypted private keys, and the size checking function should be able to return also the size of an encrypted key

Would this all make sense?

Reply:
Thinking of this, it is not very nice to put the restriction on library implementing the keystore, that the keys need to be directly accessible. I had done this for performance reasons, and for being able to generate keys...

Since access to the keys really depends on the HW and actual arch_crypto, maybe it would be better to keep this function as original (copy to buffer), and if any caching is needed for performance reasons, it could be done inside the arch_crypto